src/main/java/org/folio/circulation/resources/CheckOutByBarcodeResource.java

```diff
@@ -178,6 +198,39 @@ private CompletableFuture<Result<LoanAndRelatedRecords>> saveCheckOutSessionReco
         succeeded(records)));
   }
 
+  private CompletableFuture<Result<LoanAndRelatedRecords>> acquireLock(LoanAndRelatedRecords records, CheckOutLockRepository checkOutLockRepository, AtomicReference<String> checkOutLockId) {
+    log.debug("acquireLock:: Creating checkout lock {} ", records.getCheckoutLockConfiguration());
+    if (records.isCheckoutLockFeatureEnabled()) {
+      CompletableFuture<CheckOutLock> future = new CompletableFuture<>();
+      checkOutLockRepository.createLockWithRetry(0, future, records);
+      return future.handle((res, err) -> {
+        if (err != null) {
+          log.error("acquireLock:: Unable to acquire lock for item {} ", records.getItem().getBarcode(), err);
+          return Result.failed(ValidationErrorFailure.singleValidationError("unable to acquire lock", "", ""));
+        }
+        checkOutLockId.set(res.getId());
+        return Result.succeeded(records);
+      });
+    }
+    return completedFuture(Result.succeeded(records));
+  }
+
+  private CompletableFuture<Result<LoanAndRelatedRecords>> validateItemLimitBasedOnLockFeatureFlag(LoanAndRelatedRecords records, CheckOutValidators validators, CirculationErrorHandler errorHandler) {
+    if (records.isCheckoutLockFeatureEnabled()) {
+      return validators.refuseWhenItemLimitIsReached(Result.of(() -> records))
+        .thenApply(r -> r.next(errorHandler::failWithValidationErrors));
```

Comment:
failWithValidationErrors is a method that typically should only be called once (it is already being called from the future chain directly). The idea behind this error handler is that it accumulates validation errors (instead of failing instantly) and only fails when you call failWithValidationErrors and returns all of the accumulated errors. In this case, I'm not even sure that validation error (code 422) is appropriate.

Another question is, how is this functionality related to item limit? refuseWhenItemLimitIsReached is supposed to indicate that item limit has been reached (this limit is set by the loan policy).